hwp picoscope agent

ykyohei opened the pull request:
New agent for LC sensor in CHWP system

## Description
New agent for Picoscope  3403D MSO

## Motivation and Context
For monitoring CHWP system

## How Has This Been Tested?
Tested at UTokyo and TSAT.
Current agent is only confirmed to work by privileged container. I would like to have advices for improving this.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] Unless I am preparing a release, I have opened this PR onto the `develop` branch.

BrianJKoopman commented on socs/agent/class_ps3000a.py:
```diff
@@ -0,0 +1,519 @@
+from picosdk.ps3000a import ps3000a as ps
+from picosdk.functions import adc2mV, assert_pico_ok, splitMSODataFast
+import time
+import datetime
```
Unused import.

BrianJKoopman reviewed (commented):
Hi Kyohei, thanks for the PR. I have some inline comments, and a bunch of questions. In no particular order:

- It looks like `agents/hwp_picoscope/class_ps3000a.py` was moved into socs in one commit, but in the next commit was recommitted to the old location (in 082367aa89c9080a559472ecf0b741c32a3707de). This old copy should be removed.
- I agree it would be good to avoid the use of the privileged container. What led you to using a privileged container? What indicated things weren't working without it?
- How much data is being published to the non-downsampled feed?
- How does the rest of the HWP system interact with this data? Does anything need to react to the readings published by this Agent?
- For reading the Agent I'd find some documentation in the supporting `class_ps3000a.py` file helpful. Mostly descriptions for what the various functions that are being called are doing:
  - `pico.SigGenSingle(biasfreq)`
  - `pico.SetScopeAll()`
  - `pico.SetBufferAll()`
  - `pico.set_digital_port()`
  - `pico.set_digital_buffer()`
  - `pico.Stream_AD()`
  - `pico.close()`
- It'd also be helpful to describe the functions and expected return values for:
  - `pico.get_value()`
  - `pico.get_digital_values_simple()`
- `class_ps3000a.py` seems to borrow heavily from examples in the [picosdk repo](https://github.com/picotech/picosdk-python-wrappers/tree/master/ps3000aExamples). For instance, `SetScope()` is almost exactly [lines 38-56](https://github.com/picotech/picosdk-python-wrappers/blob/89003868b5bc52511ee57419f0afbfade25f1882/ps3000aExamples/ps3000aStreamingExample.py#L38-L56) of `ps3000aStreamingExample.py`. This is fine, I just think we need to put a copy of their license at the top of this file. You can find that [here](https://raw.githubusercontent.com/picotech/picosdk-python-wrappers/master/LICENSE.md).

I would like to understand the need for `--privileged`. We should try to understand the privileges that we do need and permit only those, rather than granting all capabilities.

ykyohei commented on agents/hwp_picoscope/Dockerfile:
```diff
@@ -0,0 +1,24 @@
+    wget https://labs.picotech.com/debian/pool/main/libp/libps3000a/libps3000a_2.1.0-6r570_amd64.deb &&\
+    dpkg -i *deb
+
+RUN git clone https://github.com/picotech/picosdk-python-wrappers.git
```
I tried to do it but If I try to install by pip, it doesn't work so I ended up installing like this.

```
picoscope_1              | Traceback (most recent call last):
picoscope_1              |   File "pico_agent.py", line 13, in <module>
picoscope_1              |     import class_ps3000a as ps
picoscope_1              |   File "/app/socs/agents/picoscope/class_ps3000a.py", line 4, in <module>
picoscope_1              |     from picosdk.ps3000a import ps3000a as ps
picoscope_1              |   File "/usr/local/lib/python3.8/dist-packages/picosdk/ps3000a.py", line 21, in <module>
picoscope_1              |     ps3000a = Ps3000alib()
picoscope_1              |   File "/usr/local/lib/python3.8/dist-packages/picosdk/ps3000a.py", line 18, in __init__
picoscope_1              |     super(Ps3000alib, self).__init__("ps3000a")
picoscope_1              |   File "/usr/local/lib/python3.8/dist-packages/picosdk/library.py", line 49, in __init__
picoscope_1              |     self._clib = self._load()
picoscope_1              |   File "/usr/local/lib/python3.8/dist-packages/picosdk/library.py", line 71, in _load
picoscope_1              |     raise CannotFindPicoSDKError("PicoSDK (%s) not found, check %s" % (self.name, env_var_name))
picoscope_1              | picosdk.errors.CannotFindPicoSDKError: PicoSDK (ps3000a) not found, check LD_LIBRARY_PATH
```

BrianJKoopman replied:
Interesting. I'm able to `import class_ps3000a` after a pip install within the container. However, if this doesn't work on your end, I'd suggest when doing the git clone, that a specific git commit is then checked out. This should ensure future builds of the image reproduce this working state, in the event the repo is updated and something breaks unexpectedly.

So I would make this:
```
RUN git clone https://github.com/picotech/picosdk-python-wrappers.git
WORKDIR /make/picosdk-python-wrappers
RUN git switch -c 89003868b5bc52511ee57419f0afbfade25f1882
RUN python3 setup.py install
```
For the most recent commit.

ykyohei replied:
Thanks for your useful suggestion!
Yes I don't get error by `import class_ps3000a`. I get this error when I do the `run_single` task,

ykyohei commented on agents/hwp_picoscope/pico_agent.py:
```diff
@@ -0,0 +1,179 @@
+
+ON_RTD = os.environ.get('READTHEDOCS') == 'True'
+if not ON_RTD:
+    import class_ps3000a as ps 
```
In class_ps3000a, this agent import picosdk . If I move class_ps3000a to socs/agent/, we cannot find the picosdk and Read the Docs build fails.
This is why I decide to protect to import class_ps3000a.
How should I fix this issue?

ykyohei replied:
In class_ps3000a, this agent import picosdk . If I move class_ps3000a to socs/agent/, we cannot find the picosdk and Read the Docs build fails.
This is why I decide to protect to import class_ps3000a.
How should I fix this issue?

ykyohei replied:
You mean I should move class_ps3000a.py to socs/agent and keep the protection?

BrianJKoopman replied:
> You mean I should move class_ps3000a.py to socs/agent and keep the protection?

Yes, that's what I had in mind. I did try to mock the import of picosdk for the documentation, but ran into trouble with it. Keeping the protection is alright.

ykyohei commented on socs/agent/class_ps3000a.py:
```diff
@@ -0,0 +1,519 @@
+        print('sample points: {}, sampling rate: {} (MHz), length: {} (sec)'.format(self.totalSamples, samplerate*1e-6, self.totalSamples*self.interval*1e-9))
+        
+        # Opens the device/s
+        self.status["openunit"] = ps.ps3000aOpenUnit(ctypes.byref(self.chandle), None)
```
Yes, in this line communication with picoscope is initialted.
Since we don't do streaming with this agent, every time when we run the `run_signle`, this agent initiate the communication with picoscope and stop the communication.
Currently this is working. Should I fix to initiate the communication only once?

BrianJKoopman replied:
I see. It's probably alright as is. How often do we expect to call the `run_single` Task?

ykyohei commented:
On using previleged containers.
Because of the lack of my understanding about the communication of picoscope, I don't know how I can reduce the privileges of the container now. I need to contact to the vendor.
Current my observations is following, please let me know if you have any suggestions or comments.
- picoscope is connected to pc by USB 3.0
- if the container is previleged, this agent works without mounting `/dev`
- If I mount `/dev`, and the container is not previleged, the agent raises an error. it cannot find picoscope. `picosdk.library.DeviceNotFoundError: Driver ps3000a could find no devices (PICO_NOT_FOUND)` I guess the replated source code is [this](https://github.com/picotech/picosdk-python-wrappers/blob/master/picosdk/library.py).

BrianJKoopman commented:
> On using previleged containers. Because of the lack of my understanding about the communication of picoscope, I don't know how I can reduce the privileges of the container now. I need to contact to the vendor.

Yeah, I'm finding it difficult to follow their code. Some thoughts below though.

> Current my observations is following, please let me know if you have any suggestions or comments.
> 
> * picoscope is connected to pc by USB 3.0
> * if the container is previleged, this agent works without mounting `/dev`
> * If I mount `/dev`, and the container is not previleged, the agent raises an error. it cannot find picoscope. `picosdk.library.DeviceNotFoundError: Driver ps3000a could find no devices (PICO_NOT_FOUND)` I guess the related source code is [this](https://github.com/picotech/picosdk-python-wrappers/blob/master/picosdk/library.py).

I think in `class_ps3000a.py`, when we call `ps.ps3000aOpenUnit()`, the arguments are `(handle, serial)`. Currently we're providing `None` for the serial. I think if this is provided, it's the serial number of the picoscope. I'm not sure providing the serial number is going to help much, as I suspect it's still going to be searching available devices, but can we try it and see if it helps?

Do you know where `self._open_unit()` comes from in their code? https://github.com/picotech/picosdk-python-wrappers/blob/89003868b5bc52511ee57419f0afbfade25f1882/picosdk/library.py#L167 I think that is likely where some search is going on for the device.

When you connect the Picoscope, does it show up in `/dev/` on the host filesystem (i.e. not in the container)? Can you see something like `/dev/ttyUSB0` show up after plugging it in? What about if you look at output from `$ dmesg` shortly after plugging in the device?

ykyohei commented:
Mm, I could not find where `self._open_unit()` comes from..

I tried `$ dmesg` and I got following output
```
[7531344.055028] usb 1-3: new high-speed USB device number 11 using xhci_hcd
[7531344.200276] usb 1-3: New USB device found, idVendor=0ce9, idProduct=1211, bcdDevice= 1.00
[7531344.200281] usb 1-3: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[7531344.200285] usb 1-3: Product: WestBridge
[7531344.200288] usb 1-3: Manufacturer: Cypress
[7531344.200291] usb 1-3: SerialNumber: 0000000004BE 
```
USB device number is different every time I plug in the picoscope.
I also checked the diff of `/dev` and I think this correponds to `/dev/bus/usb/001/011`.
Also `/dev/char/189:8` appears every time I plug in picoscope this number is also different every time.

ykyohei commented:
I tried `self.status["openunit"] = ps.ps3000aOpenUnit(ctypes.byref(self.chandle), '0000000004BE')` and this worked but actually this seem to work with `serial={any string}` as long as container is privileged.
I also tried `self.status["openunit"] = ps.ps3000aOpenUnit(ctypes.byref(self.chandle), '0000000004BE')` by non previleged container mounting `/dev`, and it didn't work.

ykyohei commented:
Another test.
I tried the former instead of the latter, but it doesn't work.
```
picoscope:
    image: simonsobs/ocs-hwp-picoscope-agent:utokyo
    hostname: ocs-nuc
    cap_add:
      - ALL
    volumes:
      - ${OCS_CONFIG_DIR}:/config:ro
      - /dev:/dev
    command:
      - "--instance-id=picoscope"
      - "--site-hub=ws://crossbar:8001/ws"
      - "--site-http=http://crossbar:8001/call"
```

```
picoscope:
    image: simonsobs/ocs-hwp-picoscope-agent:utokyo
    hostname: ocs-nuc
    privileged: true
    volumes:
      - ${OCS_CONFIG_DIR}:/config:ro
    command:
      - "--instance-id=picoscope"
      - "--site-hub=ws://crossbar:8001/ws"
      - "--site-http=http://crossbar:8001/call"
```

BrianJKoopman commented:
> Mm, I could not find where `self._open_unit()` comes from..
> 
> I tried `$ dmesg` and I got following output
> 
> ```
> [7531344.055028] usb 1-3: new high-speed USB device number 11 using xhci_hcd
> [7531344.200276] usb 1-3: New USB device found, idVendor=0ce9, idProduct=1211, bcdDevice= 1.00
> [7531344.200281] usb 1-3: New USB device strings: Mfr=1, Product=2, SerialNumber=3
> [7531344.200285] usb 1-3: Product: WestBridge
> [7531344.200288] usb 1-3: Manufacturer: Cypress
> [7531344.200291] usb 1-3: SerialNumber: 0000000004BE 
> ```
> 
> USB device number is different every time I plug in the picoscope. I also checked the diff of `/dev` and I think this correponds to `/dev/bus/usb/001/011`. Also `/dev/char/189:8` appears every time I plug in picoscope this number is also different every time.

Interesting. My usual approach here, when dealing with a device that I need to know a consistent address for is to create a udev rule to symlink it to a known location. For instance, you could create `/etc/udev/rules.d/51-picoscope.rules` with the contents:
```
SUBSYSTEM=="tty", ATTRS{idVendor}=="0ce9", ATTRS{idProduct}=="1211", ATTRS{serial}=="0000000004BE", MODE="0666", SYMLINK="picoscope"
```
And if it works you should get `/dev/picoscope`, which should be a symlink to whatever `/dev/bus/usb/001/011` the device is assigned. (This also sets the permissions on the device so that we can read/write to it, which is typically a problem with other devices by default, like the Lakeshore 240.)

However, this might only be helpful if the picoSDK took a path for the device, rather than doing the search we suspect it's doing.

BrianJKoopman commented:
> I tried `self.status["openunit"] = ps.ps3000aOpenUnit(ctypes.byref(self.chandle), '0000000004BE')` and this worked but actually this seem to work with `serial={any string}` as long as container is privileged. I also tried `self.status["openunit"] = ps.ps3000aOpenUnit(ctypes.byref(self.chandle), '0000000004BE')` by non previleged container mounting `/dev`, and it didn't work.

Strange that it works with any string as the serial. I wonder if it's just doing the search anyway.

That test of `cap_add: ALL` and mounting `/dev/` compared to `--privileged` is interesting. `--privileged` is doing a bit more than adding all capabilities and providing access to devices. I wonder what's happening with that search that is failing.

At this point I'd more lean towards running this Agent outside of a container, rather than use a container with `--privileged`.

ykyohei commented:
I made some commits to make run this agent outside of docker.
I edited the `ocs_plugin_so.py` and added the dependencies to the docs.

Also here is the errors I reproduced.
If I try to publish too much data at once, I get errors like this.
```
Recreating b499987523bb_chwp-nuc_picoscope_1 ... done
Attaching to chwp-nuc_picoscope_1

picoscope_1              | 2022-07-18T23:02:44+0000 Using OCS version 0.9.0
picoscope_1              | 2022-07-18T23:02:44+0000 ocs: starting <class 'ocs.ocs_agent.OCSAgent'> @ observatory.picoscope
picoscope_1              | 2022-07-18T23:02:44+0000 log_file is apparently None
picoscope_1              | 2022-07-18T23:02:44+0000 transport connected
picoscope_1              | 2022-07-18T23:02:44+0000 session joined:
picoscope_1              | SessionDetails(realm="test_realm",
picoscope_1              |                session=1745950184681534,
picoscope_1              |                authid="G9GW-UFNA-YY9N-EWLM-FUTM-37RX",
picoscope_1              |                authrole="iocs_agent",
picoscope_1              |                authmethod="anonymous",
picoscope_1              |                authprovider="static",
picoscope_1              |                authextra={'x_cb_node': '7632ab74a5f3-1', 'x_cb_worker': 'worker001', 'x_cb_peer': 'tcp4:10.0.1.176:32894', 'x_cb_pid': 16},
picoscope_1              |                serializer="cbor.batched",
picoscope_1              |                transport="websocket",
picoscope_1              |                resumed=None,
picoscope_1              |                resumable=None,
picoscope_1              |                resume_token=None)

picoscope_1              | 2022-07-18T23:02:56+0000 start called for run_single
picoscope_1              | 2022-07-18T23:02:56+0000 run_single:1 Status is now "starting".
picoscope_1              | 2022-07-18T23:02:56+0000 sample points: 20000, sampling rate: 3.5999999999999996 (MHz), length: 0.005540000000000001 (sec)
picoscope_1              | 2022-07-18T23:03:00+0000 Drive frequency: 150000.0 (Hz), PKtoPK: 2.0 (V)
picoscope_1              | 2022-07-18T23:03:00+0000 Capturing at sample interval 280000 ns
picoscope_1              | 2022-07-18T23:03:01+0000 Done grabbing values.
picoscope_1              | 2022-07-18T23:03:01+0000 Capturing interval 0.0056 sec
picoscope_1              | 2022-07-18T23:03:10+0000 {'sample_points': 20000, 'sampling_rate_Hz': 3600000.0, 'length_sec': 0.005540000000000001, 'Drive_frequency_Hz': 150000.0, 'PKtoPk': 2.0, 'ch_A_ave': 0.0, 'ch_B_ave': 0.4214443897637795, 'ch_C_ave': 0.9973794291338581, 'ch_D_ave': 0.6924581692913386, 'ch_A_std': 0.0, 'ch_B_std': 2.561836454135132, 'ch_C_std': 3.859541496451247, 'ch_D_std': 3.235421429389422, 'ch_B_Acos': 0.0, 'ch_C_Acos': 0.0, 'ch_D_Acos': 0.0, 'ch_B_Asin': 0.0, 'ch_C_Asin': 0.0, 'ch_D_Asin': 0.0, 'ch_0_rate': 0.0, 'ch_1_rate': 0.0, 'ch_2_rate': 0.0, 'ch_3_rate': 0.0, 'ch_4_rate': 0.0, 'ch_5_rate': 0.0, 'ch_6_rate': 0.0, 'ch_7_rate': 0.0}
picoscope_1              | 2022-07-18T23:03:12+0000 tried to send WebSocket message with size 1240381 exceeding payload limit of 1048576 octets
picoscope_1              | 2022-07-18T23:03:12+0000 Unhandled Error
picoscope_1              | Traceback (most recent call last):
picoscope_1              |   File "pico_agent.py", line 186, in main
picoscope_1              |     runner.run(agent, auto_reconnect=True)
picoscope_1              |   File "/usr/local/lib/python3.8/dist-packages/autobahn/twisted/wamp.py", line 410, in run
picoscope_1              |     reactor.run()
picoscope_1              |   File "/usr/local/lib/python3.8/dist-packages/twisted/internet/base.py", line 1318, in run
picoscope_1              |     self.mainLoop()
picoscope_1              |   File "/usr/local/lib/python3.8/dist-packages/twisted/internet/base.py", line 1328, in mainLoop
picoscope_1              |     reactorBaseSelf.runUntilCurrent()
picoscope_1              | --- <exception caught here> ---
picoscope_1              |   File "/usr/local/lib/python3.8/dist-packages/twisted/internet/base.py", line 967, in runUntilCurrent
picoscope_1              |     f(*a, **kw)
picoscope_1              |   File "/app/ocs/ocs/ocs_feed.py", line 232, in publish_message
picoscope_1              |     self.flush_buffer()
picoscope_1              |   File "/app/ocs/ocs/ocs_feed.py", line 132, in flush_buffer
picoscope_1              |     self.agent.publish(self.address,(
picoscope_1              |   File "/usr/local/lib/python3.8/dist-packages/autobahn/wamp/protocol.py", line 1476, in publish
picoscope_1              |     raise e
picoscope_1              |   File "/usr/local/lib/python3.8/dist-packages/autobahn/wamp/protocol.py", line 1472, in publish
picoscope_1              |     self._transport.send(msg)
picoscope_1              |   File "/usr/local/lib/python3.8/dist-packages/autobahn/wamp/websocket.py", line 125, in send
picoscope_1              |     self.sendMessage(payload, isBinary)
picoscope_1              |   File "/usr/local/lib/python3.8/dist-packages/autobahn/websocket/protocol.py", line 2258, in sendMessage
picoscope_1              |     raise PayloadExceededError(emsg)
picoscope_1              | autobahn.exception.PayloadExceededError: tried to send WebSocket message with size 1240381 exceeding payload limit of 1048576 octets
picoscope_1              |
picoscope_1              | 2022-07-18T23:03:12+0000 tried to send WebSocket message with size 1240381 exceeding payload limit of 1048576 octets
picoscope_1              | 2022-07-18T23:03:12+0000 Unhandled Error
picoscope_1              | Traceback (most recent call last):
picoscope_1              |   File "pico_agent.py", line 186, in main
picoscope_1              |     runner.run(agent, auto_reconnect=True)
picoscope_1              |   File "/usr/local/lib/python3.8/dist-packages/autobahn/twisted/wamp.py", line 410, in run
picoscope_1              |     reactor.run()
picoscope_1              |   File "/usr/local/lib/python3.8/dist-packages/twisted/internet/base.py", line 1318, in run
picoscope_1              |     self.mainLoop()
picoscope_1              |   File "/usr/local/lib/python3.8/dist-packages/twisted/internet/base.py", line 1328, in mainLoop
picoscope_1              |     reactorBaseSelf.runUntilCurrent()
picoscope_1              | --- <exception caught here> ---
picoscope_1              |   File "/usr/local/lib/python3.8/dist-packages/twisted/internet/base.py", line 967, in runUntilCurrent
picoscope_1              |     f(*a, **kw)
picoscope_1              |   File "/app/ocs/ocs/ocs_feed.py", line 132, in flush_buffer
picoscope_1              |     self.agent.publish(self.address,(
picoscope_1              |   File "/usr/local/lib/python3.8/dist-packages/autobahn/wamp/protocol.py", line 1476, in publish
picoscope_1              |     raise e
picoscope_1              |   File "/usr/local/lib/python3.8/dist-packages/autobahn/wamp/protocol.py", line 1472, in publish
picoscope_1              |     self._transport.send(msg)
picoscope_1              |   File "/usr/local/lib/python3.8/dist-packages/autobahn/wamp/websocket.py", line 125, in send
picoscope_1              |     self.sendMessage(payload, isBinary)
picoscope_1              |   File "/usr/local/lib/python3.8/dist-packages/autobahn/websocket/protocol.py", line 2258, in sendMessage
picoscope_1              |     raise PayloadExceededError(emsg)
picoscope_1              | autobahn.exception.PayloadExceededError: tried to send WebSocket message with size 1240381 exceeding payload limit of 1048576 octets
picoscope_1              |
picoscope_1              | 2022-07-18T23:03:12+0000 run_single:1 Single acquisition exited cleanly.
picoscope_1              | 2022-07-18T23:03:12+0000 run_single:1 Status is now "done".
```

If I split the data into smaller chunks, no errors like this.
```
picoscope_1              | 2022-07-18T23:02:46+0000 start called for run_single
picoscope_1              | 2022-07-18T23:02:46+0000 run_single:0 Status is now "starting".
picoscope_1              | 2022-07-18T23:02:46+0000 sample points: 100, sampling rate: 3.5999999999999996 (MHz), length: 2.7700000000000002e-05 (sec)
picoscope_1              | 2022-07-18T23:02:51+0000 Drive frequency: 150000.0 (Hz), PKtoPK: 2.0 (V)
picoscope_1              | 2022-07-18T23:02:51+0000 Capturing at sample interval 280000 ns
picoscope_1              | 2022-07-18T23:02:51+0000 Done grabbing values.
picoscope_1              | 2022-07-18T23:02:51+0000 Capturing interval 2.8e-05 sec
picoscope_1              | 2022-07-18T23:02:51+0000 {'sample_points': 100, 'sampling_rate_Hz': 3600000.0, 'length_sec': 2.7700000000000002e-05, 'Drive_frequency_Hz': 150000.0, 'PKtoPk': 2.0, 'ch_A_ave': 0.0, 'ch_B_ave': 0.9596456692913385, 'ch_C_ave': 1.2746062992125984, 'ch_D_ave': 0.7904773622047244, 'ch_A_std': 0.0, 'ch_B_std': 3.7983869865032704, 'ch_C_std': 4.322400259914971, 'ch_D_std': 3.445610939006991, 'ch_B_Acos': 0.0, 'ch_C_Acos': 0.0, 'ch_D_Acos': 0.0, 'ch_B_Asin': 0.0, 'ch_C_Asin': 0.0, 'ch_D_Asin': 0.0, 'ch_0_rate': 0.0, 'ch_1_rate': 0.0, 'ch_2_rate': 0.0, 'ch_3_rate': 0.0, 'ch_4_rate': 0.0, 'ch_5_rate': 0.0, 'ch_6_rate': 0.0, 'ch_7_rate': 0.0}
picoscope_1              | 2022-07-18T23:02:51+0000 run_single:0 Single acquisition exited cleanly.
picoscope_1              | 2022-07-18T23:02:51+0000 run_single:0 Status is now "done".
```

BrianJKoopman reviewed (commented):
I think this is nearly ready to merge, with the recommendation that this Agent runs directly on the host system and not in Docker. Any final additions/fixes @ykyohei?

ykyohei commented:
Thank you for your review!
Yes the current agent handles the data splitting.
I don't have any other addition/fixes now.
